src/ast/codegen_helper.cpp

```diff
@@ -0,0 +1,17 @@
+#include "codegen_helper.h"
+
+namespace bpftrace::ast {
+
+bool needAssignMapStatementAllocation(const AssignMapStatement &assignment)
+{
+  const auto &map = *assignment.map;
+  const auto &expr_type = assignment.expr->type;
+  if (shouldBeInBpfMemoryAlready(expr_type)) {
+    return !expr_type.IsSameSizeRecursive(map.type);
```

Comment:
I realized what was bothering me about this check. It only applies in the case of the resource_analyser when you're trying to get the max write value size. However when you call this from `AssignMapStatement` in codegen you shouldn't be checking this e.g. this returns false:
```
BEGIN { @a[1] = ("hellotherelongstr", 5); @a[2] = ("hellotherelongstr", 6); exit(); }
```
Further more I think some of the types in `shouldBeInBpfMemoryAlready` don't need an additional scratch buffer e.g. tuples and strings.

Reply:
Hmmm I was just copying the existing logic in CodegenLLVM::AssignMapStatement. I'm fine to change it but admittedly I don't understand the effects well.

Is there an urgency to do this in this PR? Could we do this in another PR? I'm hoping to minimize side effects

Reply:
Discussed this offline with @jordalgo - there was confusion on what this function did. Will clarify with comments why this function is extracted out and what its purpose is

Reply:
Yeah, my bad. I wasn't reading close enough.